### PR TITLE
update tf modules pinned for updates in terraform-modules

### PR DESCRIPTION
**Pull request by cmharlow**
Jira ticket: https://jira.mozilla.com/browse/SE-1919

What this PR does:
* updates terraform module pinned versions to work with updates in [mozilla-it/terraform-modules](https://github.com/mozilla-it/terraform-modules)
~* pins our usage of eks module from [mozilla-it/terraform-modules](https://github.com/mozilla-it/terraform-modules) to a commit so can avoid upstream changes altering us~

### CODE REVIEW

File: terraform/main.tf [hunk 1]
```diff
@@ -23,18 +23,19 @@ locals {
...
+      instance_types   = ["t3.large"],
       disk_size        = 100,
       subnets          = data.terraform_remote_state.vpc.outputs.public_subnets
     }
   }
 }
 
 module "itse-apps-stage-1" {
-  source                    = "github.com/mozilla-it/terraform-modules//aws/eks?ref=master"
+  source                    = "github.com/mozilla-it/terraform-modules//aws/eks?ref=d9a7662d464cd395d87fbc26a0682a50c4fb208f"
```
Comment (duallain) [inline terraform/main.tf:38]: This has been a source of contention on the team before.  Are you pinning here because you encountered a specific issue, or just as a way of trying to be correct? We own that repo that you're pointing at, so if we could 'just' take the responsibility for keeping all changes backwards compatible, which does make rolling changes out after we update tf modules quite a bit easier.

TL;DR: I could be convinced either way, but I think pointing at 'master' was the agreed upon pattern we inherited.

**Review by duallain — APPROVED** [review]
In general LGTM, one question about pinning, but merging this is OK in my opinion no matter what.

File: terraform/main.tf (hunk 1 above)
Comment (sciurus) [reply terraform/main.tf:38]: Either decision comes with tradeoffs, and our internal module versioning/publishing/pinning story is something I hope the current infrastructure working group weighs in on.

I'm okay with us using either pattern in the interim.

Comment (cmharlow) [reply terraform/main.tf:38]: so here's the reasoning behind this: some changse made in terraform-modules, particular the eks module, upstream broke both this itse-apps-stage-1 terraform and the itse-apps-prod-1 terraform. And since this is on an older version of terraform, we could use a lockfile or some other mechanism to pin down at least what had worked in the past.  So sven and myself got stuck doing some pretty nasty dependencies tracing yesterday to see what was breaking.

I've seen both patterns yall speak of - pin to master, pin to a commit, for our tf modules - in web sre stuff, and i've seen both break (e.g. alberto had pinned a module to a commit that no longer existed in another tf repo).

But the long & short of it is, these tf repos were broken yesterday by upstream changes.

Comment (cmharlow) [reply terraform/main.tf:38]: My personal preference honestly is to try to update these shared apps cluster TF repos to a later version of terraform so we can use a lockfile at least. But this is easier.

Also, I wouldn't hold my breath for the IaC group to actually decide anything at this level as ablew through out workflow considerations. So it probably behooves us to decide at least within web sre what we want to do.

Comment (cmharlow) [reply terraform/main.tf]: So reverted this to master, added a more explicit note (probably needed anyways) about the upstream changes making some of the node_group keys brittle. Let me know if other changes should be done here.

**Review by duallain — APPROVED** [review]
The note is very helpful.  Everything LGTM

File: terraform/main.tf (hunk 1 above)
Comment (smarnach) [reply terraform/main.tf]: > we could 'just' take the responsibility for keeping all changes backwards compatible

The problem is that this is impossible in general. One of the offending changes that broke this repo was [pinning the provider versions](https://github.com/mozilla-it/terraform-modules/commit/2a82b2e76f2e95d5cf52a3984080aefe9a5a0c1c#diff-a221b94eb256be38f0fd2b00504d6a866411e4adcca137bf5e0d308980d872c0). It's simply not possible to do that in a backwards-compatible way. Neither it is possible to upgrade a provider that is used in multiple places across the dependency graph in a backwards-compatible way.

Another problem is that the terraform-modules repo does not pin its external dependencies, which was another reason why it was difficult to get the Terraform code in this repo working again.

I can see that we may get to a point where we can follow the master branch of our own Terraform modules, but in the current situation I don't think this is a good idea. We definitely need to pin our external dependencies first, and in my opinion we also need to have some automation in place that monitors for differences between the Terraform code and the deployed infrastructure, becase otherwise it's just not possible to forsee the effect of changes to the terraform-modules repo.

I'm not asking for any changes here – I'm just voicing my opinion that this problem isn't solved yet, and we should solve it at some point in the future.
